# ClonEvol
Inferring and visualizing clonal evolution in multi-sample cancer sequencing.

Code & documentation is maintained at <a href="https://github.com/hdng/clonevol">https://github.com/hdng/clonevol</a>

## What is ClonEvol?
ClonEvol is a package for clonal ordering and clonal evolution visualization. It uses the clustering of heterozygous variants identified using other tools as input to infer consensus clonal evolution trees and estimate the cancer cell fraction (also called clonal frequency) of the clones in individual samples. ClonEvol can deal with statistical uncertainty and error in sequencing and data analysis that may distort the cellular prevalence estimate of individual variants.

ClonEvol is developed at <a href="http://www.maherlab.com/">Christopher Maher laboratory</a>, <a href="http://genome.wustl.edu/">The McDonnell Genome Institute</a>, <a href="http://www.wustl.edu">Washington University in St. Louis.</a>


![](images/clonevol.png)
*Fig. 1. Overview of ClonEvol. ClonEvol infers and visualizes clonal evolution trees using variants preclustered by other methods. It can also visualizes trees identified by other methods.

## How to install and use ClonEvol?

ClonEvol code and documentation is maintained at <a href="https://github.com/hdng/clonevol">https://github.com/hdng/clonevol</a>


## How to cite ClonEvol

H. X. Dang, B. S. White, S. M. Foltz, C. A. Miller, J. Luo, R. C. Fields, C. A. Maher; ClonEvol: clonal ordering and visualization in cancer sequencing, Annals of Oncology, , mdx517, https://doi.org/10.1093/annonc/mdx517

